File: directcall-java/app/src/main/java/com/sendbird/calls/quickstart/call/VideoCallActivity.java
package com.sendbird.calls.quickstart.call;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.sendbird.calls.AcceptParams;
import com.sendbird.calls.AudioDevice;
import com.sendbird.calls.CallOptions;
import com.sendbird.calls.DialParams;
import com.sendbird.calls.DirectCall;
import com.sendbird.calls.DirectCallUserRole;
import com.sendbird.calls.SendBirdCall;
import com.sendbird.calls.SendBirdVideoView;
import com.sendbird.calls.quickstart.BaseApplication;
import com.sendbird.calls.quickstart.R;
import com.sendbird.calls.quickstart.utils.ToastUtils;

import org.webrtc.RendererCommon;

import java.util.Set;

public class VideoCallActivity extends CallActivity {

    private boolean mIsVideoEnabled;

    //+ Views
    private SendBirdVideoView mVideoViewFullScreen;
    private View mViewConnectingVideoViewFullScreenFg;
    private RelativeLayout mRelativeLayoutVideoViewSmall;
    private SendBirdVideoView mVideoViewSmall;
    private ImageView mImageViewCameraSwitch;
    private ImageView mImageViewVideoOff;
    //- Views

    @Override
    protected int getLayoutResourceId() {
        return R.layout.activity_video_call;
    }

    @Override
    protected void initViews() {
        super.initViews();
        Log.i(BaseApplication.TAG, "[VideoCallActivity] initViews()");

        mVideoViewFullScreen = findViewById(R.id.video_view_fullscreen);
        mViewConnectingVideoViewFullScreenFg = findViewById(R.id.view_connecting_video_view_fullscreen_fg);
        mRelativeLayoutVideoViewSmall = findViewById(R.id.relative_layout_video_view_small);
        mVideoViewSmall = findViewById(R.id.video_view_small);
        mImageViewCameraSwitch = findViewById(R.id.image_view_camera_switch);
        mImageViewVideoOff = findViewById(R.id.image_view_video_off);
    }

    @Override
    protected void setViews() {
        super.setViews();

        mVideoViewFullScreen.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT);
        mVideoViewFullScreen.setZOrderMediaOverlay(false);
        mVideoViewFullScreen.setEnableHardwareScaler(true);

        mVideoViewSmall.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT);
        mVideoViewSmall.setZOrderMediaOverlay(true);
        mVideoViewSmall.setEnableHardwareScaler(true);

        if (mDirectCall != null) {
            if (mDirectCall.getMyRole() == DirectCallUserRole.CALLER && mState == STATE.STATE_OUTGOING) {
                mDirectCall.setLocalVideoView(mVideoViewFullScreen);
                mDirectCall.setRemoteVideoView(mVideoViewSmall);
            } else {
                mDirectCall.setLocalVideoView(mVideoViewSmall);
                mDirectCall.setRemoteVideoView(mVideoViewFullScreen);
            }
        }

        mImageViewCameraSwitch.setOnClickListener(view -> {
            if (mDirectCall != null) {
                mDirectCall.switchCamera(e -> {
                    if (e != null) {
                        Log.i(BaseApplication.TAG, "[VideoCallActivity] switchCamera(e: " + e.getMessage() + ")");
                    }
                });
            }
        });

        if (mDirectCall != null && !mDoLocalVideoStart) {
            mIsVideoEnabled = mDirectCall.isLocalVideoEnabled();
        } else {
            mIsVideoEnabled = true;
        }
        mImageViewVideoOff.setSelected(!mIsVideoEnabled);
        mImageViewVideoOff.setOnClickListener(view -> {
            if (mDirectCall != null) {
                if (mIsVideoEnabled) {
                    Log.i(BaseApplication.TAG, "[VideoCallActivity] stopVideo()");
                    mDirectCall.stopVideo();
                    mIsVideoEnabled = false;
                    mImageViewVideoOff.setSelected(true);
                } else {
                    Log.i(BaseApplication.TAG, "[VideoCallActivity] startVideo()");
                    mDirectCall.startVideo();
                    mIsVideoEnabled = true;
                    mImageViewVideoOff.setSelected(false);
                }
            }
        });

        mImageViewBluetooth.setEnabled(false);
        mImageViewBluetooth.setOnClickListener(view -> {
            mImageViewBluetooth.setSelected(!mImageViewBluetooth.isSelected());
            if (mDirectCall != null) {
                if (mImageViewBluetooth.isSelected()) {
                    mDirectCall.selectAudioDevice(AudioDevice.BLUETOOTH, e -> {
                        if (e != null) {
                            mImageViewBluetooth.setSelected(false);
                        }
                    });
                } else {
                    mDirectCall.selectAudioDevice(AudioDevice.WIRED_HEADSET, e -> {
                        if (e != null) {
                            mDirectCall.selectAudioDevice(AudioDevice.SPEAKERPHONE, null);
                        }
                    });
                }
            }
        });
    }

    protected void setLocalVideoSettings(DirectCall call) {
        mIsVideoEnabled = call.isLocalVideoEnabled();
        Log.i(BaseApplication.TAG, "[VideoCallActivity] setLocalVideoSettings() => isLocalVideoEnabled(): " + mIsVideoEnabled);
        mImageViewVideoOff.setSelected(!mIsVideoEnabled);
    }

    @Override
    protected void setAudioDevice(AudioDevice currentAudioDevice, Set<AudioDevice> availableAudioDevices) {
        if (currentAudioDevice == AudioDevice.SPEAKERPHONE) {
            mImageViewBluetooth.setSelected(false);
        } else if (currentAudioDevice == AudioDevice.BLUETOOTH) {
            mImageViewBluetooth.setSelected(true);
        }

        if (availableAudioDevices.contains(AudioDevice.BLUETOOTH)) {
            mImageViewBluetooth.setEnabled(true);
        } else if (!mImageViewBluetooth.isSelected()) {
            mImageViewBluetooth.setEnabled(false);
        }
    }

    @Override
    protected void startCall(boolean amICallee) {
        CallOptions callOptions = new CallOptions();
        callOptions.setVideoEnabled(mIsVideoEnabled).setAudioEnabled(mIsAudioEnabled);

        if (amICallee) {
            callOptions.setLocalVideoView(mVideoViewSmall).setRemoteVideoView(mVideoViewFullScreen);
        } else {
            callOptions.setLocalVideoView(mVideoViewFullScreen).setRemoteVideoView(mVideoViewSmall);
        }

        if (amICallee) {
            Log.i(BaseApplication.TAG, "[VideoCallActivity] accept()");
            if (mDirectCall != null) {
                mDirectCall.accept(new AcceptParams().setCallOptions(callOptions));
            }
        } else {
            Log.i(BaseApplication.TAG, "[VideoCallActivity] dial()");
            mDirectCall = SendBirdCall.dial(new DialParams(mCalleeIdToDial).setVideoCall(mIsVideoCall).setCallOptions(callOptions), (call, e) -> {
                if (e != null) {
                    Log.i(BaseApplication.TAG, "[VideoCallActivity] dial() => e: " + e.getMessage());
                    if (e.getMessage() != null) {
                        ToastUtils.showToast(mContext, e.getMessage());
                    }

                    finishWithEnding(e.getMessage());
                    return;
                }

                Log.i(BaseApplication.TAG, "[VideoCallActivity] dial() => OK");
                updateCallService();
            });
            setListener(mDirectCall);
        }
    }

    @SuppressLint("SourceLockedOrientationActivity")
    @TargetApi(18)
    @Override
    protected boolean setState(STATE state, DirectCall call) {
        if (!super.setState(state,call)) {
            return false;
        }

        switch (state) {
            case STATE_ACCEPTING: {
                mVideoViewFullScreen.setVisibility(View.GONE);
                mViewConnectingVideoViewFullScreenFg.setVisibility(View.GONE);
                mRelativeLayoutVideoViewSmall.setVisibility(View.GONE);
                mImageViewCameraSwitch.setVisibility(View.GONE);
                break;
            }

            case STATE_OUTGOING: {
                mVideoViewFullScreen.setVisibility(View.VISIBLE);
                mViewConnectingVideoViewFullScreenFg.setVisibility(View.VISIBLE);
                mRelativeLayoutVideoViewSmall.setVisibility(View.GONE);
                mImageViewCameraSwitch.setVisibility(View.VISIBLE);
                mImageViewVideoOff.setVisibility(View.VISIBLE);
                break;
            }

            case STATE_CONNECTED: {
                mVideoViewFullScreen.setVisibility(View.VISIBLE);
                mViewConnectingVideoViewFullScreenFg.setVisibility(View.GONE);
                mRelativeLayoutVideoViewSmall.setVisibility(View.VISIBLE);
                mImageViewCameraSwitch.setVisibility(View.VISIBLE);
                mImageViewVideoOff.setVisibility(View.VISIBLE);

                mLinearLayoutInfo.setVisibility(View.GONE);

                if (call != null && call.getMyRole() == DirectCallUserRole.CALLER) {
                    call.setLocalVideoView(mVideoViewSmall);
                    call.setRemoteVideoView(mVideoViewFullScreen);
                }
                break;
            }

            case STATE_ENDING:
            case STATE_ENDED: {
                mLinearLayoutInfo.setVisibility(View.VISIBLE);

                mVideoViewFullScreen.setVisibility(View.GONE);
                mViewConnectingVideoViewFullScreenFg.setVisibility(View.GONE);
                mRelativeLayoutVideoViewSmall.setVisibility(View.GONE);
                mImageViewCameraSwitch.setVisibility(View.GONE);
            }
            break;
        }
        return true;
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.i(BaseApplication.TAG, "[VideoCallActivity] onStart()");

        if (mDirectCall != null && mDoLocalVideoStart) {
            mDoLocalVideoStart = false;
            updateCallService();
            mDirectCall.startVideo();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.i(BaseApplication.TAG, "[VideoCallActivity] onStop()");

        if (mDirectCall != null && mDirectCall.isLocalVideoEnabled()) {
            mDirectCall.stopVideo();
            mDoLocalVideoStart = true;
            updateCallService();
        }
    }
}

File: directcall-kotlin/app/src/main/java/com/sendbird/calls/quickstart/call/VideoCallActivity.kt
package com.sendbird.calls.quickstart.call

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import com.sendbird.calls.*
import com.sendbird.calls.SendBirdCall.dial
import com.sendbird.calls.quickstart.TAG
import com.sendbird.calls.quickstart.databinding.ActivityVideoCallBinding
import com.sendbird.calls.quickstart.utils.showToast
import org.webrtc.RendererCommon

class VideoCallActivity : CallActivity() {
    private var mIsVideoEnabled = false

    private lateinit var _binding: ActivityVideoCallBinding
    override val binding: ActivityVideoCallBinding
        get() = _binding

    // region Views
    override val mLinearLayoutInfo: LinearLayout
        get() = binding.linearLayoutInfo
    override val mImageViewProfile: ImageView
        get() = binding.imageViewProfile
    override val mTextViewUserId: TextView
        get() = binding.textViewUserId
    override val mTextViewStatus: TextView
        get() = binding.textViewStatus
    override val mLinearLayoutRemoteMute: LinearLayout
        get() = binding.linearLayoutRemoteMute
    override val mTextViewRemoteMute: TextView
        get() = binding.textViewRemoteMute
    override val mRelativeLayoutRingingButtons: RelativeLayout
        get() = binding.relativeLayoutRingingButtons
    override val mImageViewDecline: ImageView
        get() = binding.imageViewDecline
    override val mLinearLayoutConnectingButtons: LinearLayout
        get() = binding.linearLayoutConnectingButtons
    override val mImageViewAudioOff: ImageView
        get() = binding.imageViewAudioOff
    override val mImageViewBluetooth: ImageView
        get() = binding.imageViewBluetooth
    override val mImageViewEnd: ImageView
        get() = binding.imageViewEnd
    // endregion

    override fun onCreate(savedInstanceState: Bundle?) {
        _binding = ActivityVideoCallBinding.inflate(layoutInflater)
        super.onCreate(savedInstanceState)
    }

    override fun setViews() {
        super.setViews()
        binding.videoViewFullscreen.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT)
        binding.videoViewFullscreen.setZOrderMediaOverlay(false)
        binding.videoViewFullscreen.setEnableHardwareScaler(true)
        binding.videoViewSmall.setScalingType(RendererCommon.ScalingType.SCALE_ASPECT_FIT)
        binding.videoViewSmall.setZOrderMediaOverlay(true)
        binding.videoViewSmall.setEnableHardwareScaler(true)
        mDirectCall?.let { mDirectCall ->
            if (mDirectCall.myRole == DirectCallUserRole.CALLER && mState === STATE.STATE_OUTGOING) {
                mDirectCall.setLocalVideoView(binding.videoViewFullscreen)
                mDirectCall.setRemoteVideoView(binding.videoViewSmall)
            } else {
                mDirectCall.setLocalVideoView(binding.videoViewSmall)
                mDirectCall.setRemoteVideoView(binding.videoViewFullscreen)
            }
        }
        binding.imageViewCameraSwitch.setOnClickListener {
            mDirectCall?.switchCamera { e ->
                if (e != null) {
                    Log.i(TAG, "[VideoCallActivity] switchCamera(e: " + e.message + ")")
                }
            }
        }
        mIsVideoEnabled = if (mDirectCall != null && !mDoLocalVideoStart) {
            mDirectCall?.isLocalVideoEnabled ?: false
        } else {
            true
        }
        binding.imageViewVideoOff.isSelected = !mIsVideoEnabled
        binding.imageViewVideoOff.setOnClickListener {
            mDirectCall ?: return@setOnClickListener
            if (mIsVideoEnabled) {
                Log.i(TAG, "[VideoCallActivity] stopVideo()")
                mDirectCall?.stopVideo()
                mIsVideoEnabled = false
                binding.imageViewVideoOff.isSelected = true
            } else {
                Log.i(TAG, "[VideoCallActivity] startVideo()")
                mDirectCall?.startVideo()
                mIsVideoEnabled = true
                binding.imageViewVideoOff.isSelected = false
            }
        }
        mImageViewBluetooth.isEnabled = false
        mImageViewBluetooth.setOnClickListener {
            mDirectCall ?: return@setOnClickListener
            mImageViewBluetooth.isSelected = !mImageViewBluetooth.isSelected
            if (mImageViewBluetooth.isSelected) {
                mDirectCall?.selectAudioDevice(AudioDevice.BLUETOOTH) { e ->
                    if (e != null) {
                        mImageViewBluetooth.isSelected = false
                    }
                }
            } else {
                mDirectCall?.selectAudioDevice(AudioDevice.WIRED_HEADSET) { e ->
                    if (e != null) {
                        mDirectCall?.selectAudioDevice(AudioDevice.SPEAKERPHONE, null)
                    }
                }
            }
        }
    }

    fun setLocalVideoSettings(call: DirectCall) {
        mIsVideoEnabled = call.isLocalVideoEnabled
        Log.i(TAG,"[VideoCallActivity] setLocalVideoSettings() => isLocalVideoEnabled(): $mIsVideoEnabled")
        binding.imageViewVideoOff.isSelected = !mIsVideoEnabled
    }

    override fun setAudioDevice(currentAudioDevice: AudioDevice?, availableAudioDevices: Set<AudioDevice>?) {
        if (currentAudioDevice == AudioDevice.SPEAKERPHONE) {
            mImageViewBluetooth.isSelected = false
        } else if (currentAudioDevice == AudioDevice.BLUETOOTH) {
            mImageViewBluetooth.isSelected = true
        }
        if (availableAudioDevices != null && availableAudioDevices.contains(AudioDevice.BLUETOOTH)) {
            mImageViewBluetooth.isEnabled = true
        } else if (!mImageViewBluetooth.isSelected) {
            mImageViewBluetooth.isEnabled = false
        }
    }

    override fun startCall(amICallee: Boolean) {
        val callOptions = CallOptions().setVideoEnabled(mIsVideoEnabled).setAudioEnabled(mIsAudioEnabled).apply {
            if (amICallee) {
                setLocalVideoView(binding.videoViewSmall).setRemoteVideoView(binding.videoViewFullscreen)
            } else {
                setLocalVideoView(binding.videoViewFullscreen).setRemoteVideoView(binding.videoViewSmall)
            }
        }
        if (amICallee) {
            Log.i(TAG, "[VideoCallActivity] accept()")
            mDirectCall?.accept(AcceptParams().setCallOptions(callOptions))
        } else {
            Log.i(TAG, "[VideoCallActivity] dial()")
            val mCalleeIdToDial = mCalleeIdToDial ?: return
            mDirectCall = dial(DialParams(mCalleeIdToDial).setVideoCall(mIsVideoCall).setCallOptions(callOptions)) { _, e ->
                if (e != null) {
                    Log.i(TAG, "[VideoCallActivity] dial() => e: " + e.message)
                    if (e.message != null) {
                        showToast(e.message ?: "")
                    }
                    finishWithEnding(e.message ?: "")
                    return@dial
                }
                Log.i(TAG, "[VideoCallActivity] dial() => OK")
                updateCallService()
            }
            setListener(mDirectCall)
        }
    }

    @SuppressLint("SourceLockedOrientationActivity")
    override fun setState(state: STATE?, call: DirectCall?): Boolean {
        if (!super.setState(state, call)) {
            return false
        }
        when (state) {
            STATE.STATE_ACCEPTING -> {
                binding.videoViewFullscreen.visibility = View.GONE
                binding.viewConnectingVideoViewFullscreenFg.visibility = View.GONE
                binding.relativeLayoutVideoViewSmall.visibility = View.GONE
                binding.imageViewCameraSwitch.visibility = View.GONE
            }
            STATE.STATE_OUTGOING -> {
                binding.videoViewFullscreen.visibility = View.VISIBLE
                binding.viewConnectingVideoViewFullscreenFg.visibility = View.VISIBLE
                binding.relativeLayoutVideoViewSmall.visibility = View.GONE
                binding.imageViewCameraSwitch.visibility = View.VISIBLE
                binding.imageViewVideoOff.visibility = View.VISIBLE
            }
            STATE.STATE_CONNECTED -> {
                binding.videoViewFullscreen.visibility = View.VISIBLE
                binding.viewConnectingVideoViewFullscreenFg.visibility = View.GONE
                binding.relativeLayoutVideoViewSmall.visibility = View.VISIBLE
                binding.imageViewCameraSwitch.visibility = View.VISIBLE
                binding.imageViewVideoOff.visibility = View.VISIBLE
                mLinearLayoutInfo.visibility = View.GONE
                if (call != null && call.myRole == DirectCallUserRole.CALLER) {
                    call.setLocalVideoView(binding.videoViewSmall)
                    call.setRemoteVideoView(binding.videoViewFullscreen)
                }
            }
            STATE.STATE_ENDING, STATE.STATE_ENDED -> {
                mLinearLayoutInfo.visibility = View.VISIBLE
                binding.videoViewFullscreen.visibility = View.GONE
                binding.viewConnectingVideoViewFullscreenFg.visibility = View.GONE
                binding.relativeLayoutVideoViewSmall.visibility = View.GONE
                binding.imageViewCameraSwitch.visibility = View.GONE
            }
            else -> {}
        }
        return true
    }

    override fun onStart() {
        super.onStart()
        Log.i(TAG, "[VideoCallActivity] onStart()")
        if (mDirectCall != null && mDoLocalVideoStart) {
            mDoLocalVideoStart = false
            updateCallService()
            mDirectCall?.startVideo()
        }
    }

    override fun onStop() {
        super.onStop()
        Log.i(TAG, "[VideoCallActivity] onStop()")
        if (mDirectCall != null && mDirectCall?.isLocalVideoEnabled == true) {
            mDirectCall?.stopVideo()
            mDoLocalVideoStart = true
            updateCallService()
        }
    }
}